jetty-server/src/main/java/org/eclipse/jetty/server/PlainOrSslConnectionFactory.java

```diff
@@ -0,0 +1,192 @@
+//  ------------------------------------------------------------------------
+//  All rights reserved. This program and the accompanying materials
+//  are made available under the terms of the Eclipse Public License v1.0
+//  and Apache License v2.0 which accompanies this distribution.
+//
+//      The Eclipse Public License is available at
+//      http://www.eclipse.org/legal/epl-v10.html
+//
+//      The Apache License v2.0 is available at
+//      http://www.opensource.org/licenses/apache2.0.php
+//
+//  You may elect to redistribute this code under either of these licenses.
+//  ========================================================================
+//
+
+package org.eclipse.jetty.server;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+
+import org.eclipse.jetty.io.AbstractConnection;
+import org.eclipse.jetty.io.Connection;
+import org.eclipse.jetty.io.EndPoint;
+import org.eclipse.jetty.util.BufferUtil;
+import org.eclipse.jetty.util.log.Log;
+import org.eclipse.jetty.util.log.Logger;
+
+/**
+ * <p>A ConnectionFactory whose connections detect whether the first bytes are
+ * TLS bytes and upgrades to either a TLS connection or to a plain connection.</p>
+ */
+public class PlainOrSslConnectionFactory extends AbstractConnectionFactory
+{
+    private static final Logger LOG = Log.getLogger(PlainOrSslConnection.class);
+    private static final int TLS_ALERT_FRAME_TYPE = 0x15;
+    private static final int TLS_HANDSHAKE_FRAME_TYPE = 0x16;
+    private static final int TLS_MAJOR_VERSION = 3;
+
+    private final SslConnectionFactory sslConnectionFactory;
+    private final String plainProtocol;
+
+    /**
+     * <p>Creates a new plain or TLS ConnectionFactory.</p>
+     * <p>If {@code plainProtocol} is {@code null}, and the first bytes are not TLS, then
+     * {@link #unknownProtocol(ByteBuffer, EndPoint)} is called; applications may override its
+     * behavior (by default it closes the EndPoint) for example by writing a minimal response. </p>
+     *
+     * @param sslConnectionFactory The SslConnectionFactory to use if the first bytes are TLS
+     * @param plainProtocol        the protocol of the ConnectionFactory to use if the first bytes are not TLS, or null.
+     */
+    public PlainOrSslConnectionFactory(SslConnectionFactory sslConnectionFactory, String plainProtocol)
+    {
+        super("plain|ssl");
+        this.sslConnectionFactory = sslConnectionFactory;
+        this.plainProtocol = plainProtocol;
+    }
+
+    @Override
+    public Connection newConnection(Connector connector, EndPoint endPoint)
+    {
+        return configure(new PlainOrSslConnection(endPoint, connector), connector, endPoint);
+    }
+
+    /**
+     * @param buffer The buffer with the first bytes of the connection
+     * @return whether the bytes seem TLS bytes
+     */
+    protected boolean seemsTLS(ByteBuffer buffer)
+    {
+        int tlsFrameType = buffer.get(0) & 0xFF;
+        int tlsMajorVersion = buffer.get(1) & 0xFF;
```

Comment:
How does this work with these bytes being consumed?  Shouldn't they be peeked ?

Reply:
The bytes are not consumed because I use `get(int)` so the `position` is never changed, so `get(int)` is a peek indeed.